### ISSUE TITLE
Wrong longitude in some PISCES variables

### ISSUE THREAD

**Issue by klauswyser** [EC-Earth/ece2cmor3#837]
After processing OptimESM experiments with ece2cmor3, the cmorfixer raises an error with lon/lat, e.g.:
````
2024-10-30 15:51:57 INFO:cmor-fixer.py: Replacing the longitude and latitude t-g
rid vertices for vertices_longitude (none) in /nobackup/rossby27/proj/optimesm/c
morized/CD43-1/CMIP6/CMIP/EC-Earth-Consortium/EC-Earth3-ESM-1/abrupt-4xCO2/r1i1p
1f1/Omon/chldiatos/gn/v20240925/chldiatos_Omon_EC-Earth3-ESM-1_abrupt-4xCO2_r1i1
p1f1_gn_195701-195712.nc
````
It does so only for t-grid, and only for these variables: 
````
o2os_Omon
chldiatos_Omon
limfediat_Omon
sios_Omon
chlos_Omon
limirrdiat_Omon
limirrmisc_Omon
chlmiscos_Omon
limfemisc_Omon
po4os_Omon
no3os_Omon
dfeos_Omon
limnmisc_Omon
limndiat_Omon
````
They seem to be all from PISCES, or?

When inspecting one of the files we find that the longitude in the original file looks totally weird (left):

<img src="https://github.com/user-attachments/assets/3c220a48-d271-4b43-9cd0-8ec56f8a54de" width=45%\>
<img src="https://github.com/user-attachments/assets/d1838e39-0914-426d-adce-437995435bcb" width=45%\>

After running cmorfixer the longitude looks fine (right). There are also some minor corrections in the latitude, but that seems to be more kind of rounding errors.

Have you seen this before? For OptimESM or for CMIP6 datasets?

Why does it occur only with PISCES, and not NEMO (ocean) or LIM variables?

Why does it happen only for these 14 variables but not for all PISCES variables on the t-grid?

Most important: can this been fixed somehow?

And yes, we find this in all experiments and all years that we have checked, so it's not just a randomly occuring glitch in the post-processing.

**Comment by etiennesky** [comment]
I am not sure we had this issue with CMIP6 as we were using a slightly different output control file.

**Comment by klauswyser** [comment]
Just found that all these variables are saved in the pisces_grid_T_SFC file. nav_lon and nav_lat look fine in this nemo output file, so something in ece2cmor3 must go wrong.

**Comment by etiennesky** [comment]
I think the pisces_grid_T_SFC is the same as grid_T which has 3 poles and repeated lines, so maybe ece2cmor3 needs to account for that, as it does for the grid_T ?

**Comment by vlap** [comment]
Hi!

I have never seen the pattern plotted above, nor am I aware of any special issues with pisces_grid_T_SFC. 

> Most important: can this been fixed somehow?

Replacing lon in those files with good values for 2D grid_T should do the job. With nco in a big loop. But I understand that doing this without  finding out the root of the problem will not feel nice.

**Comment by goord** [comment]
The grid identification is going wrong. ece2cmor3 thinks these variables are living on the w-grid. I will fix it asap

**Comment by goord** [comment]
The issue appears to be the grid identification expecting _grid_T_2D instead of _grid_T_SFC. I made a fix in a new branch sfc-grid-fix, ready for testing

**Comment by goord** [comment]
Hi @etiennesky when I look in the history, I conclude that the issue must have been there from the start. Are you sure the pisces files were named _grid_T_SFC in the old runs and not _grid_T_2D?

All in all, the way we map files to grids is a bit error-prone and unsustainable in my opinion. Probably, we should throw an error if the grid cannot be identified, or better, assume it's a new grid altogether. In the latter case we will have 2 copies of the same grid which can be slightly slower and use more memory, but at least the output will contain the correct lats/lons....

**Comment by etiennesky** [comment]
well, this was never detected by us and we ran hundreds of experiments.

on my system I can see files produced for OptimESM by Valentina on hpc2020 using a recent version of ece2cmor3 and they are fine, I checked o2os_Omon, dfeos_Omon and fgco2_Omon.

**Comment by goord** [comment]
Hi @etiennesky , can you send the raw pisces output file and necessary metadata files for the OpimESM cmorization for those variables?

**Comment by klauswyser** [comment]
@etiennesky, what did you check in your files? The files can be opened and everything looks fine. Did you compare longitude_vertices of one of these PISCES files against say Amon.tos? The data is fine, and so are longitude and latitude, it's the vertices that are different and  trigger the cmorfixer.

Actually there is a twist to this: the cmorfixer sets new vertices, however, these some of these new vertices are actually degenerate. Not sure where the vertices that the cmorfixer imposes  come from.

**Comment by klauswyser** [comment]
Thanks @goord, your sfc-grid-fix branch seems to fix the problem.

**Comment by treerink** [comment]
> @etiennesky, what did you check in your files? The files can be opened and everything looks fine. Did you compare longitude_vertices of one of these PISCES files against say Amon.tos? The data is fine, and so are longitude and latitude, it's the vertices that are different and trigger the cmorfixer.
> 
> Actually there is a twist to this: the cmorfixer sets new vertices, however, these some of these new vertices are actually degenerate. Not sure where the vertices that the cmorfixer imposes come from.

The vertices are from the files in the dir `cmor-fixer/nemo-vertices`. See https://github.com/EC-Earth/cmor-fixer/issues/27.

**Comment by treerink** [comment]
I think the `sfc-grid-fix` branch can be merged, right?

**Comment by klauswyser** [comment]
Agree. I have tested ece2mor with the sfc-grid-fix changes, then checked the results with cmorfixer. It looks allright, no more errors about vertices.

**Comment by etiennesky** [comment]
ah ok sorry I missed the problem, I though the longitude was wrong (by reading your comment "When inspecting one of the files we find that the longitude in the original file looks totally weird (left):"), but you mean to say the longitude vertices are wrong...

**Comment by klauswyser** [comment]
The problem popped up when plotting the longitude with panoply which is why I wrote the above text. But you are right, it probably is rather a problem of vertices_longitude.

I change the title of the issue to avoid confusion.

**Comment by etiennesky** [comment]
> The longitde_vertices error may have gone unnoticed. Cmorisation with ece2cmor3 seemingly worked fine, no error. The error is raised by cmorfixer, but it's not fatal. So if you haven't applied cmorfixer on your data or just applied it without bothering about the logfiles you wouldn't have noticed any problem.

that's a big problem with ece2cmor (and cmorfixer) is that errors are non-fatal so are mostly never spotted, can we do something about this?

**Comment by goord** [comment]
Hi @etiennesky I agree with you, the coding style and language are forgiving and can lead to undetected mistakes. In this case the detection of the grid is necessary to load the correct vertices from a file and we try to deduce the grid from the nemo output file name. But, there are all sorts of file names; for instance, we need to insert the T-grid vertices for file ending in
`
lim_grid_T_2D
lim_grid_W_3D
lim_grid_T_3D_ncatice
opa_zoom_700_sum
pisces_grid_T_SFC
`
but not for
`
opa_grid_ptr_T_3basin_2D
`
The lack of a systematic file naming has lead us to create a fuzzy logic scheme, which may silently assign the wrong vertices for unseen file names. We can only fix this by having systematic file naming convention. Or maybe there is some metadata in the files themselves that would allow us to unambiguously detect the grid?

**Comment by klauswyser** [comment]
Well, isn't it that anything which is not grid_U or grid_V is grid_T? Mabe we should mke the logic around this instead.

Another and really failproof idea is to use the lon/lat of the first gridpoint to identify the grid.

And finally, the NEMO filenames are not set in stone but set in the XIOS file_def.xml file, We could make sure that all NEMO output file names contain the proper grid id. This is not very practical for ECE3 because so many datasets have already been produced, but for ECE4 we need a general overview of the cmorisation anyway and _if_ we decide to  use ece2cmor3 we could make this a clear requirement for NEMO output files from ECE4.

**Comment by etiennesky** [comment]
> Agree. I have tested ece2mor with the sfc-grid-fix changes, then checked the results with cmorfixer. It looks allright, no more errors about vertices.

Hi @klauswyser just to be sure I understand the resolution, can the cmor fixer be used on the data that was produced with a previous version of ece2cmor3?

**Comment by treerink** [comment]
@etiennesky Yes `cmor-fixer` can be used for any EC-Earth3 output  which is cmorised by any ece2cmor3 version (say above v1.2). The cmor-fixer is robust against already fixed data either by an earlier call by cmor-fixer or because a meanwhile corrected ece2cmor3 version. I released this week a new `cmor-fixer` release `4.0`, but with respect to this the same applies to `cmor-fixer` `v3.0`.

**Comment by klauswyser** [comment]
@etiennesky : yes, you can apply the cmorfixer and it fixes the corner lon/lat. I had an older version of the cmorfixer and a newer ece2cmor3, yet the cmorfixer still detected the problem with the corners and fixed it.
